sstables/compress.cc

```diff
@@ -335,39 +335,59 @@ compression::locate(uint64_t position, const compression::segmented_offsets::acc
 
 }
 
-template <ChecksumUtils ChecksumType>
+// For SSTables 2.x (formats 'ka' and 'la'), the full checksum is a combination of checksums of compressed chunks.
+// For SSTables 3.x (format 'mc'), however, it is supposed to contain the full checksum of the file written so
+// the per-chunk checksums also count.
+enum class compressed_checksum_mode {
+    checksum_chunks_only,
+    checksum_all,
+};
+
+template <ChecksumUtils ChecksumType, compressed_checksum_mode mode>
 class compressed_file_data_source_impl : public data_source_impl {
     std::optional<input_stream<char>> _input_stream;
     sstables::compression* _compression_metadata;
     sstables::compression::segmented_offsets::accessor _offsets;
     sstables::local_compression _compression;
+    std::optional<uint32_t> _expected_digest;
+    uint32_t _actual_digest;
```

Comment:
It's overkill, but we can omit these data members when unneeded. Wrap them in a template struct that is empty for checksum_chunks_only and contains these members for checksum_all and add it as a member (with [[no_unique_address]]).

Reply:
Alternatively: `std::conditional_t<..., uint32_t, struct {} [[no_unique_address]]> _actual_digest;`

Reply:
`checksum_chunks_only` and `checksum_all` define how the digest was calculated (depends on SSTable format), not if digest will be checked.